agent/background.go

```diff
@@ -30,11 +28,13 @@ func (a *Agent) startHeartbeat(ctx context.Context, cancel context.CancelFunc, t
 			signalBeat, err = a.doHeartbeat(ctx, tc)
 			if signalBeat == evergreen.TaskConflict {
 				tc.logger.Task().Error("Encountered task conflict while checking heartbeat, aborting task")
-				heartbeat <- evergreen.TaskFailed
+				if err != nil {
```

Comment:
So this was the unreachable code path, correct? Is there a reason why we no longer want to assign something to heartbeat?

Reply:
The [select that waits for the heartbeat](https://github.com/evergreen-ci/evergreen/blob/a4cd5cd31a93e48486fd275876723b565008a653/agent/agent.go#L520-L527) can accept either context cancellation or a heartbeat status, so from a correctness/consistency perspective, the heartbeat must either cancel the context _or_ send a task status, but not both. In this case, I think `evergreen.StatusConflict` is not a real task status, so it seems like cancelling is preferable to sending a fake task status and hoping that it correctly handles it.

Reply:
It isn't that we are no longer sending a status to the heartbeat channel - it was actually never set up this way, here is an older version of the same function: https://github.com/evergreen-ci/evergreen/blob/42e23e63cb85a0f303e7e7dbfa36f37b3b85c556/agent/background.go#L32

In hindsight this cancel() was never getting reached though.  In that regard we are making it such that the old code is reachable.  

A child context is created from the parent context used in the heartbeat loop [here](https://github.com/evergreen-ci/evergreen/blob/main/agent/agent.go#L498): When the parent heartbeat context gets cancelled, this innerContext  Done channel is closed and it looks like the two startTask and handleTaskResponse functions will not complete, causing the agent to return back to its high-level loop looking for new tasks which I think is the intended scenario when we run into a task conflict.

Conversely when we receive an abort signal we have historically just been sending a failed signal through the heartbeat channel which propagates down through the agent to [finishTask](https://github.com/evergreen-ci/evergreen/blob/098d5c237fde79ce9cf7a6884b8362f332c512c7/agent/agent.go#L577), and will cause post task commands etc to run. I think it makes sense that we wouldn't want to cancel the context when this happens which is why we saw some problems with the change deployed yesterday that cancelled context on abort signals.